demos/jobs/jobs_demo_mosquitto/jobs_demo_mosquitto.c

```diff
@@ -0,0 +1,1264 @@
+ * @param[in] jobid the job ID
+ * @param[in] jobidLength size of the job ID string
+ * @param[in] report body of the update
+ *
+ * @return true if libmosquitto accepted the publish message;
+ * false otherwise
+ *
+ * @note This does not call mosquitto_loop(); it expects main() to do so.
+ */
+static bool sendUpdate( handle_t * h,
+                        char * jobid,
+                        size_t jobidLength,
+                        char * report );
+
+/**
+ * @brief Read job ID and URL from a JSON job document
+ *
+ * @param[in] h runtime state handle
+ * @param[in] message an MQTT publish message
+ *
+ * @return true if values were found and copied to the handle;
+ * false otherwise
+ */
+static bool parseJob( handle_t * h,
+                      const struct mosquitto_message * message );
+
+/**
+ * @brief libmosquitto callback for a received publish message
+ *
+ * @param[in] m unused
+ * @param[in] p runtime state handle
+ * @param[in] message an MQTT publish message
+ *
+ * This checks if a message corresponds to a Jobs API, and transitions
+ * runtime state based on the API and current state.
+ */
+void on_message( struct mosquitto * m,
+                 void * p,
+                 const struct mosquitto_message * message );
+
+/**
+ * @brief publish a request to the Jobs service to start a job
+ *
+ * @param[in] h runtime state handle
+ *
+ * @return true if libmosquitto accepted the publish message;
+ * false otherwise
+ *
+ * @note This does not call mosquitto_loop(); it expects main() to do so.
+ */
+static bool sendStartNext( handle_t * h );
+
+/**
+ * @brief reports status of the download process
+ *
+ * @param[in] h runtime state handle
+ *
+ * @return the result of sendUpdate()
+ */
+static bool update( handle_t * h );
+
+/**
+ * @brief launch a download process
+ *
+ * @param[in] h runtime state handle
+ *
+ * @return true if fork() succeeded;
+ * false otherwise
+ */
+static bool download( handle_t * h );
+
+/**
+ * @brief kill a download process
+ *
+ * @param[in] h runtime state handle
+ */
+static void cancelDownload( handle_t * h );
+
+/**
+ * @brief libmosquitto callback for log messages
+ *
+ * @param[in] m unused
+ * @param[in] p unused
+ * @param[in] level unused
+ * @param[in] log the message to print
+ */
+static void on_log( struct mosquitto * m,
+                    void * p,
+                    int level,
+                    const char * log );
+
+/**
+ * @brief generic signal handler
+ *
+ * @param[in] signal the caught signal value
+ */
+static void catch( int signal );
+
+/**
+ * @brief setup signal handling and libmosquitto
+ *
+ * @param[in] h runtime state handle
+ *
+ * @return false if a step failed;
+ * true otherwise
+ */
+static bool setup( handle_t * h );
+
+/**
+ * @brief disconnect and clean up
+ *
+ * @param[in] x unused
+ * @param[in] p runtime state handle
+ */
+static void teardown( int x,
+                      void * p );
+
+/**
+ * @brief log an informational message
```

Comment:
Should `warnx` be used for _informational_ messages? Wouldn't the user expectation be that non-error messages are printed to `stdout`?

Reply:
changed

Reply:
I changed it back.  The stderr stream has the benefit of being unbuffered, and I don't want to complicate the program with setvbuf().  The functions I'm using for messages like err() and warn() have the very nice property that they include strerror() output.